File: app/src/main/java/ro/noadsx15/hybrid/DnsPacket.java
package ro.noadsx15.hybrid;

import java.nio.charset.StandardCharsets;
import java.util.*;

public final class DnsPacket {
 private DnsPacket(){}
 public static String host(byte[] q){if(q.length<17)return null;int i=12;List<String> labels=new ArrayList<>();while(i<q.length){int n=q[i++]&255;if(n==0)break;if(n>63||i+n>q.length)return null;labels.add(new String(q,i,n,StandardCharsets.US_ASCII));i+=n;}return labels.isEmpty()?null:String.join(".",labels);}
 public static int questionEnd(byte[]q){int i=12;while(i<q.length){int n=q[i++]&255;if(n==0)return Math.min(i+4,q.length);if(n>63||i+n>q.length)return-1;i+=n;}return-1;}
 public static byte[] nxdomain(byte[]q){int e=questionEnd(q);byte[]r=Arrays.copyOf(q,e>0?e:q.length);if(r.length>=12){r[2]=(byte)0x81;r[3]=(byte)0x83;r[4]=0;r[5]=1;for(int i=6;i<=11;i++)r[i]=0;}return r;}
}

File: app/src/main/java/ro/noadsx15/app/DnsPacket.kt
package ro.noadsx15.app

object DnsPacket {
    fun host(query: ByteArray): String? {
        if (query.size < 17) return null
        var i = 12; val labels = ArrayList<String>()
        while (i < query.size) {
            val length = query[i++].toInt() and 255
            if (length == 0) break
            if (length > 63 || i + length > query.size) return null
            labels.add(String(query, i, length, Charsets.US_ASCII)); i += length
        }
        return labels.joinToString(".").takeIf { it.isNotEmpty() }
    }

    fun questionEnd(query: ByteArray): Int {
        var i=12
        while(i<query.size){ val n=query[i++].toInt() and 255; if(n==0)return (i+4).coerceAtMost(query.size); if(n>63||i+n>query.size)return -1;i+=n }
        return -1
    }

    fun nxdomain(query: ByteArray): ByteArray {
        val end=questionEnd(query); val response=query.copyOf(if(end>0)end else query.size)
        if(response.size>=12){response[2]=0x81.toByte();response[3]=0x83.toByte();response[4]=0;response[5]=1;for(i in 6..11)response[i]=0}
        return response
    }
}
